core/src/com/unciv/logic/automation/unit/UnitAutomation.kt

```diff
@@ -276,34 +278,56 @@ object UnitAutomation {
 
-        if (unit.baseUnit.isAirUnit()) {
-            return false
-        }
 
         val unitDistanceToTiles = unit.movement.getDistanceToTiles()
-        val swapableTiles = unitDistanceToTiles.keys.filter { it.militaryUnit != null && it.militaryUnit!!.owner == unit.owner}.reversed()
-        for (swapTile in swapableTiles) {
-            val otherUnit = swapTile.militaryUnit!!
+        val closestCity = unit.civ.cities.minByOrNull { it.getCenterTile().aerialDistanceTo(unit.getTile()) }?.takeIf { it.getCenterTile().aerialDistanceTo(unit.getTile()) < 20 }
+
+        // Finding the distance to the closest enemy is expensive, so lets sort the tiles using a cheaper function
+        val sortedTilesToRetreatTo: Sequence<Tile> = if (closestCity != null) {
+            // If we have a city, lets favor the tiles closer to that city
+            unitDistanceToTiles.asSequence().map { it.key }.sortedBy { it.aerialDistanceTo(closestCity.getCenterTile()) }
+        } else {
+            // Rare case, what if we don't have a city nearby?
+            // Lets favor the tiles that don't have enemies close by
+            // Ideally we should check in a greater radius but might get way too expensive
+            unitDistanceToTiles.asSequence().map { it.key }.sortedByDescending { unit.civ.threatManager.getDistanceToClosestEnemyUnit(it, 3, false) }
+        }
+
+        // Lets check all tiles and swap with the first one
+        for (retreatTile in sortedTilesToRetreatTo) {
             val ourDistanceToClosestEnemy = unit.civ.threatManager.getDistanceToClosestEnemyUnit(unit.getTile(),6, false)
-            if (otherUnit.health > 80
-                && ourDistanceToClosestEnemy < otherUnit.civ.threatManager.getDistanceToClosestEnemyUnit(otherUnit.getTile(),6,false)) {
-                if (otherUnit.baseUnit.isRanged()) {
-                    // Don't swap ranged units closer than they have to be
-                    val range = otherUnit.baseUnit.range
-                    if (ourDistanceToClosestEnemy < range)
-                        continue
-                }
-                if (unit.movement.canUnitSwapTo(swapTile)) {
-                    unit.movement.swapMoveToTile(swapTile)
+            val tileDistanceToClosestEnemy = unit.civ.threatManager.getDistanceToClosestEnemyUnit(retreatTile,6,false)
+            if (ourDistanceToClosestEnemy >= tileDistanceToClosestEnemy) continue
+
+            if (retreatTile.militaryUnit == null) {
+                // See if we can retreat to the tile
+                if (unit.movement.canMoveTo(retreatTile)) {
+                    unit.movement.moveToTile(retreatTile)
                     return true
                 }
+            } else if (retreatTile.militaryUnit!!.civ == unit.civ) {
+                // The tile is taken, lets see if we want to swap retreat to it
+                val otherUnit = retreatTile.militaryUnit!!
```

Comment:
This can be moved above the if/else (without the !! at the end) so we can check nullability on it and use if not null